### PR TITLE
3D View upgrade and fixes: Visualizing the world imagery in 3D View

### PR DESCRIPTION
**Pull request by omid-esrafilian**
The following changes have been made to the 3D View:

- World imagery corresponding to the .osm file is also visualized in the 3D View along with the buildings (the imagery data changes based on the map provider specified in the Application Settings -> Maps).
- The RTL and the waypoints sequence number are added to the 3D View
- The MapScale widget is not visible while the 3D View is open
- qmldirs for Viewer3D are moved to the src/QmlControls directory. All the Viewer3D modules can now be recognized by Qt Creator.
- Rearranged the qml files/folders and CMakeList for each directory is added.
- Fixed issue regarding Ardupilot for visualizing the mission

An example of the new changes is illustrated in the figure below:

Screenshot from 2024-03-07 22-55-36

### CODE REVIEW

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
I don't see the changes I added to the code. 
Please note that the code has not been merged yet into **"master"**, so if you wish to have these changes, please make sure that you are using the branch **"3D_View_with_imagery "** of my GitHub repository.

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
It is the same as building the QGC code. Please follow the following steps:

1) git clone --recursive -j8 https://github.com/omid-esrafilian/qgroundcontrol-viewer3D.git
2) cd qgroundcontrol-viewer3D
4) git checkout 3D_View_with_imagery
3) git submodule update --recursive

Then you can open and build the project using Qt Creator. Then it should compile it.

**Comment by SebastianViollaz** [comment]
I find this:

Here an overview of the mission:
image

**Null value in item** 
image
I see a null value in the Take Off item

image

It seems like the "L" is showed too but a null value too.

**No map loaded**
As you can see in the previews photos the map isn't load properly. IN the 2D view is OK but no in 3D.
I change the source from Bing to Google and the problem solved. I don known if the problem is the source or what because in the 2D view is properly showed.
image


**ROI not showed**
As you can see in this image the ROI is no longer visible in the 3D view. 
image
image



This is the flight plan that i'm using so you can test it
[LaVerbena_V2.zip](https://github.com/mavlink/qgroundcontrol/files/14637661/LaVerbena_V2.zip)

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
Thanks much for the feedback.
Can you please also share with me the osm file you use, then I can test the whole scenarios and find out what is the problem.

**Comment by SebastianViollaz** [comment]
[LaVervena.zip](https://github.com/mavlink/qgroundcontrol/files/14638176/LaVervena.zip)
This is the OSM file

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
And which flight control do you use?

**Comment by SebastianViollaz** [comment]
Im testing with the SITL Ardupilot sim (master version) [https://github.com/ArduPilot/ardupilot](url). (Arducopter)
I'm in linux 22.04LTS (That could be the problem)
I have a non sim Arducopter, i let you known any news when i test it with this feature.

**Comment by SebastianViollaz** [comment]
I think that in the mission that  i provied i don use the ROI that is in the tool side bar in the plan editor. 
I add a mission item that change it command to the ROI (That because with the ROI that is in the tools bar an error is given). THis error was recently fixed in the master version #11159 if you marge maybe you can solve the problem of the ROI not showed

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
Thanks for the information.
I have tested the code with PX4 flight control. But, in order to display the mission in 3D View, I use the MAVLINK command type to differentiate each item in the mission. I assumed that the items in the mission were independent of the flight control system being used. However, it seems that they are not. Therefore, please let me test the code with Ardupilot to confirm this.

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
The problem was that Ardupilot uses "MAV_CMD_DO_SET_ROI" to set the ROI which is Deprecated. 
I have also included the latest changes from the upstream master branch. 
Additionally, you should be able to use Bing as the map provider now.
Can you please test the code and see if the issues are solved?

**Comment by SebastianViollaz** [comment]
Tested, properly working for me

**Comment by omid-esrafilian** [comment]
@SebastianViollaz 
Perfect. Thanks much for the feedback.

**Review by DonLakeFlyer — CHANGES_REQUESTED** [review]

File: src/ui/preferences/FlyViewSettings.qml [hunk 1]
```diff
@@ -253,13 +253,20 @@ SettingsPage {
                     text:       qsTr("Select File")
 
                     onClicked: {
+                        var filename = qsTr(_viewer3DOsmFilePath.rawValue);
```
[the remark this answers is not in the document]
Comment (omid-esrafilian) [reply src/ui/preferences/FlyViewSettings.qml]: @DonLakeFlyer 
Thanks for your comment. I just fixed that.

**Comment by DonLakeFlyer** [comment]
@HTRamsey Can you look at the docs failure? I Quick Look by me seems to indicate that the check for a push has been removed and things which should be push only are running on pulls.

**Comment by omid-esrafilian** [comment]
@DonLakeFlyer 
The CI builds seem to work without error now, after rebasing with upstream/master.

**Pull request merged**